### ISSUE TITLE
Precompile statement emitter sometimes emits invalid statements

### ISSUE THREAD

**Issue by KristofferC** [JuliaLang/julia#28808]
This is OK:

```
julia> typeof(dateformat"YYYY-mm-dd\THH:MM:SS")
DateFormat{Symbol("YYYY-mm-dd\\THH:MM:SS"),
```

But running with `trace-compile=stderr` we can see the following precompile statement being generated

```
julia> Dates.format(now(), dateformat"YYYY-mm-dd\THH:MM:SS")
precompile(Tuple{typeof(Dates.format), Dates.DateTime, Dates.DateFormat{Symbol("YYYY-mm-dd\THH:MM:SS") ...
```

which contains an unescaped `\`. in the `DateFormat` type. Encountering such a type in precompilation gathering means it will fail.

**Comment by tehrengruber** [comment]
As already mentioned in #32574 I also stumbled upon this recently. As of now I am aware of three cases in which the output is not correct.

- Types with symbols as parameters which need to be escaped, e.g. contain a backslash (this is the case you mention)
 _Example_:
  ```julia
  foo(::T) where T = 1
  struct A{B} end
  foo(A{Symbol("a\\b")})
  ```
  _Output_:
  ```
  precompile(Tuple{typeof(Main.foo), Type{Main.A{Symbol("a\b")}}})
  ```
  I have already fixed this case in https://github.com/tehrengruber/julia/commit/0fed122e57cbbd6e0bb3a7e6e56b1ab95420d10c (will open a pull request at some point)
- Types with symbols as parameters generated via `gensym`
  _Example_:
  ```
  macro add_parametric_type(expr)
    push!(expr.args[2].args, gensym())
    esc(expr)
  end
  
  @add_parametric_type struct A{} end
  
  foo(::T) where T = 1
  
  foo(A)
  ```
  _Output_:
  ```
  precompile(Tuple{typeof(Main.foo), Type{Main.A{363} where 363}})
  ```
- Types with a `Char` as a parameter, e.g. `Char(0x56)`
  _Example_:
  ```
  foo(::T) where T = 1
  struct A{B} end
  foo(A{Char(0x56)})
  ```
  _Output_:
  ```
  precompile(Tuple{typeof(Main.foo), Type{Main.A{Char(0x56000000)}}})
  ```
  The origin of the problem is in `jl_static_show_x_` here https://github.com/JuliaLang/julia/blob/faefe2ae64fb9a12824832582e2ddd3bc554d372/src/rtutils.c#L986 The bitstype `Char` (4 bytes) is traversed bytewise in reverse order, hence resulting in `0x56000000` instead of `0x00000056`. I have no clue whats the reason for this ordering. @JeffBezanson you wrote this piece of code, any reason for the ordering that I'm overlooking here?

**Comment by pfitzseb** [comment]
We also sometimes generate code with invalid `_`s in it (in type parameters, IIRC). 
See [here](https://github.com/JuliaLang/PackageCompiler.jl/blob/a403d8f3c69870e24cd41cc746bef4e926c5b055/src/snooping.jl#L77-L80) for PackageCompilers workaround.

**Comment by rai-nhdaly** [comment]
Regarding `Case 2`, copying over this context from the original issue (https://github.com/JuliaLang/julia/issues/32574):

> I would be willing to fix this myself, but I'm not sure how to handle this correctly. The generated symbol can not be used directly since the resulting code would be invalid just as well and replacing it by a different "regular" symbol, e.g. `A{##363} where ##363` by `A{B} where B`, is problematic since `B` might collide with other parameters in general.

This is now addressable thanks to the cool `var""` syntax that @c42f added in https://github.com/JuliaLang/julia/pull/32408.

So i think we just need to find whatever code is dumping the `precompile` statement (is it `jl_static_show_x_`?) and have it use the `var""` syntax here instead! :)

This would be a very similar exercise to that taken in https://github.com/JuliaLang/julia/pull/34888.

**Comment by NHDaly** [comment]
I'm working on addressing `Case 2`, here: https://github.com/JuliaLang/julia/pull/38049

It's fixed on that branch, and now i'm just following up with a couple other small improvements to `static_show()` that I noticed along the way.

**Comment by NHDaly** [comment]
Via the logging added to PackageCompiler in https://github.com/JuliaLang/PackageCompiler.jl/pull/457, I've identified two more classes of failures. One of them might just need to be fixed in PackageCompiler, but I think the other one needs to be fixed in julia:

### Case 4 - Incorrectly emitting `precompile()` with abstract types

For some reason, out of the 15627 statements emitted in our last run, we noticed 21 statements that returned `false` when running them in the output-o process. From what I can tell, these all appear to be cases where for whatever reason Julia emitted the precompile statement with **abstract types,** even though the function is _not_ marked `@nospecialize`. Sometimes they're `::Any`, and sometimes they're other abstract types or type unions. This is not a valid precompile statement, since you cannot produce machine code for a function specialized to abstract types (at least as far as I understand).

Here are some examples we see:
```julia
┌ Debug: Precompilation failed: precompile(Tuple{typeof(Base.Broadcast.broadcasted), Function, Function, Array{MathOptInterface.ScalarAffineTerm{Float64}, 1}})
└ @ Main.anonymous none:33
```
```julia
┌ Debug: Precompilation failed: precompile(Tuple{typeof(Base.Docs.docstr), Any, Any})
└ @ Main.anonymous none:33
```
```julia
┌ Debug: Precompilation failed: precompile(Tuple{typeof(Base.deepcopy_internal), Any, Base.IdDict{Any, Any}})
└ @ Main.anonymous none:33
```
```julia
┌ Debug: Precompilation failed: precompile(Tuple{typeof(DelveSDK._convert_input_value_type), Type{T} where T})
└ @ Main.anonymous none:33
```
```julia
┌ Debug: Precompilation failed: precompile(Tuple{typeof(DelveSDK._convert_input_value_type), Type{T} where T})
└ @ Main.anonymous none:33
```
```julia
┌ Debug: Precompilation failed: precompile(Tuple{typeof(Blobs.self_size), Type{Blobs.Blob{T}}} where T)
└ @ Main.anonymous none:33
```

And indeed, the above statements don't precompile, for example:
```julia
julia> precompile(Tuple{typeof(Base.Docs.docstr), Any, Any})
false
```

I'm not sure yet why julia emits statements that will fail to precompile, but this is another case that we should try to fix.



### Case 5 - `UndefVarError` caused by unknown Module names.

I _think_ this one is probably just an issue in PackageCompiler where we're not re-loading all of the modules from the user's program correctly?

In our latest run producing 15627 statements, we had 292 `UndefVarError`s that all appear to be related to module names not being available. Here are some examples:
```julia
┌ Debug: Error executing precompile(Tuple{Type{Base.Pair{A, B} where B where A}, Symbol, ColorTypes.RGBA{Float64}}):
│ LoadError("string", 1, UndefVarError(:ColorTypes))
└ @ Main.anonymous none:38
```
```julia
┌ Debug: Error executing precompile(Tuple{Type{Base.Pair{A, B} where B where A}, Symbol, Plots.Plot{Plots.GRBackend}}):
│ LoadError("string", 1, UndefVarError(:Plots))
└ @ Main.anonymous none:38
```
```julia
┌ Debug: Error executing precompile(Tuple{getfield(DelveBenchmarkRunner.ML, Symbol("#57#60"))}):
│ LoadError("string", 1, UndefVarError(:DelveBenchmarkRunner))
└ @ Main.anonymous none:38
```
```julia
┌ Debug: Error executing precompile(Tuple{Type{Measures.Length{:pct, Float64}}, Float64}):
│ LoadError("string", 1, UndefVarError(:Measures))
└ @ Main.anonymous none:38
```

I'm not sure why these ones failed when most of the statements include a module like this, and succeed. So there's something weird going on there. But _probably_ this is something internal to PackageCompiler?

**Comment by timholy** [comment]
For your case 4, these are likely due to partial inference (e.g., from being called by a `@nspecialize` method or outright inference failure) or module-level `@nospecialize`. For example, `Tuple{typeof(Base.Docs.docstr), Any, Any}`  comes from [this `@nospecialize`](https://github.com/JuliaLang/julia/blob/8c11d3c89ec400df4884cecfa784322039c10b34/base/docs/Docs.jl#L14) which applies to the entire `Docs` module.

**Comment by NHDaly** [comment]
Oh _interesting_, thanks! 

> But I do think Julia should not complain about precompiling that signature.

Yeah, agreed! this seems like it might be a bug. I've checked, and these statements are emitted _again_ when re-running my snoop file with the resulting sysimg, so I do not believe they are being statically compiled.

Thanks for the context, @timholy

**Comment by NHDaly** [comment]
Yeah, indeed I think `Case 4` is a bug in julia. I've opened an issue for it here:
https://github.com/JuliaLang/julia/issues/38149

**Comment by NHDaly** [comment]
(okay and indeed Case 5 was related to PackageCompiler; i've opened a PR here: https://github.com/JuliaLang/PackageCompiler.jl/pull/463)

**Comment by NHDaly** [comment]
Alright, and finally `Class 6` -- this is the last of the errors we're currently seeing in our usage, I think, although it's also the biggest:

### Class 6: `precompile` _succeeds_, but somehow isn't serialized during static compilation.

This last class of issue is for statements that appear to `precompile()` successfully, yet doing so during sysimg construction _does not_ get cached as you would expect. Here is an example:
```julia
precompile(Tuple{typeof(Base.push!), Array{Function, 1}, Function})
```
This function shows up in the `--trace-compile` output both during snoop compiling and again when using the newly created sysimg:

```julia
julia> using PackageCompiler

julia> mktemp() do f, io
           write(io, "precompile(Tuple{typeof(Base.push!), Array{Function, 1}, Function})")
           close(io)
           PackageCompiler.create_sysimage(sysimage_path="./precompiled", precompile_statements_file=f)
       end
[ Info: PackageCompiler: creating system image object file, this might take a while...
julia>
```
```bash
$ julia --trace-compile=stderr -J./precompiled -e 'precompile(Tuple{typeof(Base.push!), Array{Function, 1}, Function})'
precompile(Tuple{typeof(Base.MainInclude.include), String})
precompile(Tuple{typeof(Base.push!), Array{Function, 1}, Function})
```
Compare that with another function that _does_ cache its precompiled result correctly:
```julia
precompile(Tuple{typeof(Base.:(&)), Int64, Int64})
```
as seen here:
```julia
julia> mktemp() do f, io
           write(io, """
               precompile(Tuple{typeof(Base.:(&)), Int64, Int64})                   # works as expected
               precompile(Tuple{typeof(Base.push!), Array{Function, 1}, Function})  # doesn't cache
               """)
           close(io)
           PackageCompiler.create_sysimage(sysimage_path="./precompiled", precompile_statements_file=f)
       end
[ Info: PackageCompiler: creating system image object file, this might take a while...

julia>
```
```bash
julia -J./precompiled --trace-compile=stderr -e '
               precompile(Tuple{typeof(Base.:(&)), Int64, Int64})                   # works as expected
               precompile(Tuple{typeof(Base.push!), Array{Function, 1}, Function})  # doesnt cache
'
precompile(Tuple{typeof(Base.MainInclude.include), String})
precompile(Tuple{typeof(Base.push!), Array{Function, 1}, Function})
```

Out of the 15627 precompile statements emitted, we had ** 2035 cases** of this! In total, including all of the above cases, we have 2629 failures/errors, and in practice we find that only about 50-60% of the compilation time is recovered by using a statically compiled binary. I am assuming that this last class of failures is the bulk of the problem, but I don't yet know why.

**Comment by KristofferC** [comment]
Could Case 6 be from invalidations or type piracy?

Have you tried the "double sysimage" trick (https://github.com/JuliaLang/PackageCompiler.jl/issues/429). Basically you compile a sysimage, and then using that sysimage as the base, you compile the sysimage again. At that point, all invalidations should have been resolved.

**Comment by NHDaly** [comment]
Oh, wow! Thanks for the link; I hadn't seen that. That's very fascinating.

Could it explain a very simple case like the one I show above, where we're literally just precompiling the one single function?

Is the idea that i should then run PackageCompiler _starting from_ the sysimg and compile a new sysimg?

**Comment by KristofferC** [comment]
> Is the idea that i should then run PackageCompiler starting from the sysimg and compile a new sysimg?

Yes, and use `incremental=true`. Note that this is just a theory but it seems this has helped other people, so worth trying.

**Comment by NHDaly** [comment]
Hrmm no dice. :(

I ran `julia -J./precompiled`, then re-did the `create_sysimage()` from above, but this time I added `incremental=true`, and still the new process prints the same trace-compile for only `precompile(Tuple{typeof(Base.push!), Array{Function, 1}, Function})`. :(

Thanks for the info though; i'll look into applying this to our build overall. (though currently static compilation takes about an hour, so running it twice is a tough pill to swallow... 😬)

**Comment by mlubin** [comment]
This is also an issue for printf statements. As @KristofferC helped me debug on slack:
```julia
julia> precompile(Tuple{typeof(Printf.format), Base.TTY, Printf.Format{Base.CodeUnits{UInt8, String}, Tuple{Printf.Spec{Base.Val{Char(0x67000000)}}, Printf.Spec{Base.Val{Char(0x67000000)}}, Printf.Spec{Base.Val{Char(0x67000000)}}, Printf.Spec{Base.Val{Char(0x67000000)}}}}, Float64, Float64, Vararg{Float64, N} where N})
ERROR: Base.CodePointError{UInt32}(0x67000000)
Stacktrace:
 [1] code_point_err(u::UInt32)
   @ Base ./char.jl:86
 [2] Char(u::UInt32)
   @ Base ./char.jl:160
 [3] top-level scope
   @ REPL[3]:1
```

**Comment by KristofferC** [comment]
I think that is Case 3 in https://github.com/JuliaLang/julia/issues/28808#issuecomment-511200909.

**Comment by IanButterworth** [comment]
I think this is happening on master currently?
```
Generating REPL precompile statements... 36/36
┌ Warning: Failed to precompile expression
│   form = Tuple{typeof(Base.trypoptask), Base.InvasiveLinkedListSynchronized{Task}})
│   exception = Base.Meta.ParseError("invalid character \"\" near column 1")
└ @ nothing nothing:0
┌ Warning: Failed to precompile expression
│   form = precompile(Tuple{typeof(Base.print), Base.GenericIOBuffer{Array{UInt8, 1}}, UInt32})
│   exception = Base.Meta.ParseError("invalid character \"\" near column 1")
└ @ nothing nothing:0
┌ Warning: Failed to precompile expression
│   form = precompile(Tuple{typeof(Base.task_done_hook), Task})
│   exception = Base.Meta.ParseError("invalid character \"\" near column 1")
└ @ nothing nothing:0
┌ Warning: Failed to precompile expression
│   form = precompile(Tuple{typeof(Base.trypoptask), Base.InvasiveLinkedListSynchronized{Task}})
│   exception = Base.Meta.ParseError("invalid character \"\" near column 1")
└ @ nothing nothing:0
┌ Warning: Failed to precompile expression
│   form =  Base.Invastypeof(Base.trypoptask), Base.InvasiveLinkedListSynchronized{Task}})
│   exception = Base.Meta.ParseError("extra token \"}\" after end of expression")
└ @ nothing nothing:0
Executing precompile statements... 1299/133
```

**Comment by vtjnash** [comment]
That looks like something is concurrently corrupting your tmpfiles while executing them

**Comment by IanButterworth** [comment]
That happening on both my ubuntu and macos setup. Two separate machines. Is it not happening for other people?

**Comment by IanButterworth** [comment]
> Is it not happening for other people?

@ararslan reported on slack to also be seeing it on MacOS and Ubuntu. Different failures on each

**Comment by DilumAluthge** [comment]
Did it start recently? If so, a bisect could be useful

**Comment by Moelf** [comment]
```
Generating REPL precompile statements... 36/36
┌ Warning: Failed to precompile expression
│   form = precompile(Tuple{typeof(Base.print), Base.GenericIOBuffer{Array{UInt8, 1}}, String})
│   exception = Base.Meta.ParseError("invalid character \"\" near column 1")
└ @ nothing nothing:0
┌ Warning: Failed to precompile expression
│   form = .trypoptask), Base.InvasiveLinkedListSynchronized{Task}})
│   exception = Base.Meta.ParseError("invalid character \"\" near column 1")
└ @ nothing nothing:0
```

on Arch, with a80afe5524d3e3d6aac021e1d6618f130d73c776

**Comment by IanButterworth** [comment]
A bisect indicated the latest issue started at 3f6abcf77ff33420edee8ceae88970671e74c8ec which was part of https://github.com/JuliaLang/julia/pull/43415 
cc. @JeffBezanson

**Comment by IanButterworth** [comment]
A few things appear wrong here with the error message too.
```
exception = Base.Meta.ParseError("invalid character \"\" near column 1")
```
It seems the `"` escaping isn't working and it appears that it's trying to show an invisible char, but not hitting the invisible char error branch here that might give more helpful info

https://github.com/JuliaLang/julia/blob/0a91e719ea4b3c085602c8d194ee708392d27328/src/julia-parser.scm#L596-L597

**Comment by IanButterworth** [comment]
The latest issue here is fixed by making the sysimage generation single-threaded https://github.com/JuliaLang/julia/pull/44281

**Comment by Moelf** [comment]
isn't `sysimage` generation always single-threaded which is why it's so slow? or I'am mis-remembering.

**Comment by IanButterworth** [comment]
True. Just clarified my comment

**Comment by senhalil** [comment]
Not sure if what I am observing is exactly due to this issue but I am having a similar behavior to what NHDaly describes [here](https://github.com/JuliaLang/julia/issues/28808#issuecomment-715559325).

Basically I create a system image with a `precompile_statements_file` that includes the following statements but running with `--trace-compile` gives back the same statements:
```
precompile(Tuple{typeof(ProtoBuf.defaultval), Type})
precompile(Tuple{typeof(ProtoBuf.read_field), Base.IOStream, proto.v1beta1.RunRequest, ProtoBuf.ProtoMetaAttribs{Any}, UInt64, Type})
precompile(Tuple{typeof(ProtoBuf.read_field), Base.GenericIOBuffer{Array{UInt8, 1}}, proto.v1beta1.RunMetadata, ProtoBuf.ProtoMetaAttribs{Any}, UInt64, Type})
precompile(Tuple{typeof(ProtoBuf.read_field), Base.GenericIOBuffer{Array{UInt8, 1}}, proto.v1beta1.RunSpecification, ProtoBuf.ProtoMetaAttribs{Any}, UInt64, Type})
```

Tried with the [double system image trick](https://github.com/JuliaLang/julia/issues/28808#issuecomment-715540917) with no avail. It happens both on 1.8.5 and 1.9.0-rc1.

@KristofferC is there a way to check if this is due to invalidations or type piracy as you mentioned [here](https://github.com/JuliaLang/julia/issues/28808#issuecomment-715540917)

@NHDaly did you managed to find a solution to this issue?

**Comment by KristofferC** [comment]
Doesn't look like those statements have invalid syntax so seems like a different issue.

**Comment by NHDaly** [comment]
Thanks for all the work in this thread, everyone! I'm glad to see this feature has improved dramatically in robustness.

Closing this ticket in favor of the new, consolidated list, here:
https://github.com/JuliaLang/julia/issues/58756#issuecomment-2980871184
